src/search/airportsearch.cpp

```diff
@@ -744,14 +750,166 @@ void AirportSearch::randomFlightplanClicked()
+        indexDeparture = QRandomGenerator::global()->bounded(countResult);
+        do
+        {
+          departureSuccess = false;
+          if(triedIndexDeparture.count() == countResult)
+            goto allDeparturesTried;
+          while(triedIndexDeparture.contains(indexDeparture))
+          {
+            indexDeparture = (indexDeparture + 1) % countResult;
+          }
+          triedIndexDeparture.insert(indexDeparture, true);
+          departureSuccess = true;
+        }
+        while(data[indexDeparture].second.getLonX() == atools::geo::Pos::INVALID_VALUE || data[indexDeparture].second.getLatY() == atools::geo::Pos::INVALID_VALUE);
+      }
+
+      QMap<int, bool> triedIndexDestination;                                      // acts as a lookup which indices have been tried already; QMap keys are sorted, lookup is very fast
 
-  // Fetch data from SQL model
-  QVector<std::pair<int, atools::geo::Pos> > result;
-  controller->getSqlModel()->getFullResultSet(result);
+      triedIndexDestination.insert(indexDeparture, true);                         // destination shall != departure
 
-  qDebug() << Q_FUNC_INFO << "found" << result.size();
+      bool destinationSuccess;
 
+      const double lon1 = data[indexDeparture].second.getLonX() * degToRad;
+      const double lat1 = data[indexDeparture].second.getLatY() * degToRad;
+      double dist;
+
+      do
+      {
+        destinationSuccess = false;
+        float lonX, latY;
+        do
+        {
+          if(triedIndexDestination.count() == countResult)
+            goto destinationsEnd;                                                 // all destinations have been depleted, try a different departure
+          if(triedIndexDestination.count() < randomLimit)                         // the "if" is inside the "while" for the destination because the indices get used up during "picking"
+          {
+            do
+            {
+              indexDestination = QRandomGenerator::global()->bounded(countResult);
+            }
+            while(triedIndexDestination.contains(indexDestination));
+          }
+          else
+          {
+            indexDestination = QRandomGenerator::global()->bounded(countResult);
+            while(triedIndexDestination.contains(indexDestination))
+            {
+              indexDestination = (indexDestination + 1) % countResult;
+            }
+          }
+          triedIndexDestination.insert(indexDestination, true);
+          lonX = data[indexDestination].second.getLonX();
+          latY = data[indexDestination].second.getLatY();
+        }
+        while(lonX == atools::geo::Pos::INVALID_VALUE || latY == atools::geo::Pos::INVALID_VALUE);
+        const double lon2 = lonX * degToRad;
+        const double lat2 = latY * degToRad;
+        dist = R_earth * qAcos(qSin(lat1) * qSin(lat2) + qCos(lat1) * qCos(lat2) * qCos(lon2 - lon1));     // http://www.movable-type.co.uk/scripts/latlong.html Spherical Law of Cosines
```

Comment:
or use `atools::geo::Pos::distanceMeterTo()` :wink:

Reply:
yes, i changed it.